brainframe/api/bf_codecs/alarm_codecs.py

```diff
@@ -69,45 +70,44 @@ def from_dict(d):
                          stream_id=d["stream_id"])
 
 
+@dataclass
 class Alert(Codec):
     """This is sent when an Alarm has been triggered."""
 
-    def __init__(self, *,
-                 alarm_id: int,
-                 zone_id: int,
-                 stream_id: int,
-                 start_time: float,
-                 end_time: float,
-                 verified_as: Optional[bool],
-                 id_: int = None):
-        self.id: int = id_
-        """A unique identifier"""
-        self.alarm_id: int = alarm_id
-        """The ID of the alarm that this alert came from"""
-        self.zone_id: int = zone_id
-        """The ID of the zone this alert pertains to"""
-        self.stream_id: int = stream_id
-        """The ID of the stream this alert pertains to"""
-        self.start_time: float = start_time
-        """When the event started happening, in Unix Time (seconds)"""
-        self.end_time: Optional[float] = end_time
-        """When the event stopped happening, in Unix Time (seconds), or None
-        if the alert is ongoing.
-        """
-        self.verified_as: Optional[bool] = verified_as
-        """
-        - If True, then this alert was labeled by a person as legitimate
-        - If False, then this alert was labeled by a person as a false alarm
-        - If None, then this alert has not been reviewed by a person
-        """
+    alarm_id: int
+    """The ID of the alarm that this alert came from"""
+
+    zone_id: int
+    """The ID of the zone this alert pertains to"""
+
+    stream_id: int
+    """The ID of the stream this alert pertains to"""
+
+    start_time: float
+    """When the event started happening, in Unix Time (seconds)"""
+
+    end_time: Optional[float]
+    """When the event stopped happening, in Unix Time (seconds), or None
+    if the alert is ongoing.
+    """
+
+    verified_as: Optional[bool]
+    """
+    - If True, then this alert was labeled by a person as legitimate
+    - If False, then this alert was labeled by a person as a false alarm
+    - If None, then this alert has not been reviewed by a person
+    """
+
+    id: Optional[int] = None
+    """A unique identifier"""
 
     def to_dict(self):
```

Comment:
Bryce said:

> can we inherit from a class/use a metaclass that provides the to_dict method?

Reply:
So, taking a look, we have 5 codecs that just run `self.__dict__` and 13 codecs that have to do something else either in addition to or instead of running `self.__dict__`.

I wonder though if we could make `Codec.to_dict` smarter to cut down on some boilerplate :thinking:. Here are the needs of every codec's `to_dict` method (some are in more than one category):

| Description                                                                      | Classes                                                                                                                |
|----------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------|
| Uses `self.__dict__` only                                                        | Alert, Attribute, Encoding, Identity, Premises                                                                         |
| Has embedded codecs that need to have `to_dict` called on them                   | ZoneAlarmCountCondition, ZoneAlarmRateCondition, Detection, Plugin, ZoneStatus                                         |
| Has a list or dict of embedded codecs that need to have `to_dict` called on them | ZoneAlarm, Detection, LicenseInfo, Plugin, Zone, ZoneStatus                                                            |
| Has enums that need to have `value` called on them                               | StreamConfiguration, ZoneAlarmCountCondition, ZoneAlarmRateCondition, LicenseInfo, NodeDescription, PluginOption, User |
| Has a UUID object that needs to be converted to a string                         | Detection                                                                                                              |
| Has a date that needs to be converted to a string                                | LicenseTerms                                                                                                           

So, if we were to add the following simple rules, we could make most of these automatic:

- If a field `isinstance(my_field, Codec)`, we call `.to_dict()` on it
- If a field `isinstance(my_field, Enum)`, we call `.value` on it
- If a field `isinstance(my_field, UUID)`, we call `str()` on it
- If a field `isinstance(my_field, datetime.date)`, we call `.isoformat()` on it

If we add recursive rules, we can make all of these automatic:

- If a field `isinstance(my_field, list)`, we recursively call these checks on each element
- If a field `isinstance(my_field, dict)`, we recursively call these checks on each value

I would be curious about if you and @apockill think this is a good idea or introduces too much complexity.

Reply:
1) very nicely formatted 👀
2) These are good ideas! It might be worth experimenting, but I wouldn't want it to hold things up if there turn out to be a lot of corner cases. If that's the case, we might want to scope it for the next release.

Reply:
Okay, maybe it would be best to split that up into _another_ PR so that we can judge it on its own merits.

Reply:
why don't we use
http://docs.peewee-orm.com/en/latest/peewee/playhouse.html#model_to_dict

Reply:
wait nvm these are the codecs, not the models